src/audio/module_adapter/module/intel_modules.c

```diff
@@ -0,0 +1,329 @@
+	return init_internal(mod_adp);
+}
+
+static int adp_wrapper_prepare(struct processing_module *mod)
+{
+	struct ModuleAdapter* mod_adp = (struct ModuleAdapter* ) mod->priv.module_adapter;
+	return prepare_internal(mod_adp);
+}
+
+static int adp_wrapper_set_processing_mode(struct processing_module *mod,
+				enum module_processing_mode mode)
+{
+	struct ModuleAdapter* mod_adp = (struct ModuleAdapter* ) mod->priv.module_adapter;
+	set_processing_mode_internal(mod_adp, mode);
+
+	return 0;
+}
+
+static enum module_processing_mode adp_wrapper_get_processing_mode(struct processing_module *mod)
+{
+	struct ModuleAdapter* mod_adp = (struct ModuleAdapter* ) mod->priv.module_adapter;
+	return (enum module_processing_mode)get_processing_mode_internal(mod_adp);
+}
+
+static int adp_wrapper_reset(struct processing_module *mod)
+{
+	struct ModuleAdapter* mod_adp = (struct ModuleAdapter* ) mod->priv.module_adapter;
+	reset_internal(mod_adp);
+	return 0;
+}
+
+
+static int adp_wrapper_free(struct processing_module *mod)
+{
+	struct ModuleAdapter* mod_adp = (struct ModuleAdapter* ) mod->priv.module_adapter;
+	return free_internal(mod_adp);
+}
+
+
+static int adp_wrapper_process(struct processing_module *mod,
+		struct input_stream_buffer *input_buffers,
+		int num_input_buffers, struct output_stream_buffer *output_buffers,
+		int num_output_buffers)
+{
+	struct ModuleAdapter* mod_adp = (struct ModuleAdapter* ) mod->priv.module_adapter;
+	return process_internal(mod_adp, input_buffers, num_input_buffers,
+				output_buffers, num_output_buffers);
+}
+
+
+static int adp_wrapper_set_configuration(struct processing_module *mod,
+		 uint32_t config_id,
+		 enum module_cfg_fragment_position pos, uint32_t data_offset_size,
+		 const uint8_t *fragment, size_t fragment_size, uint8_t *response,
+		 size_t response_size)
+{
+	struct ModuleAdapter* mod_adp = (struct ModuleAdapter* ) mod->priv.module_adapter;
+	return set_configuration_internal(mod_adp, config_id, pos, data_offset_size, fragment,
+			fragment_size, response, response_size);
+}
+
+/* Processing Module Adapter API*/
+static struct module_interface adp_interface = {
+	.init  = intel_modules_init,
+	.prepare = intel_modules_prepare,
+	.process = intel_modules_process,
+	.set_processing_mode = adp_wrapper_set_processing_mode,
+	.get_processing_mode = adp_wrapper_get_processing_mode,
+	.set_configuration = adp_wrapper_set_configuration,
+	.reset = adp_wrapper_reset,
+	.free = intel_modules_free,
+};
+
+/**
+ * \brief Init comp_driver data.
+ * \param[in] drv - component driver pointer.
+ *
+ * \note: For dynamically loaded module the comp_driver data are not yet initialized.
+ *        They need to be set in runtime, since base FW get uuid and module informations
+ *        during load time.
+ */
+static void intel_modules_init_drv(struct comp_driver *drv)
+{
+	drv->ops.create2 = *intel_modules_shim_new;
+	drv->ops.prepare = module_adapter_prepare;
```

Comment:
create2? That is interesting.

Reply:
Yes, this is consequence of limitation of original drv->ops.create() method. The original comp_ops definition looks as follows:
struct comp_dev *(*create)(const struct comp_driver *drv,
				   struct comp_ipc_config *ipc_config,
				   void *ipc_specific_config)

ip_spec_config is casted in each component device implementation into its proprietary configuration structure, known at the compilation time. This is because modules code is linked with base firmware. 
In this particular case we don't know how looks the configuration structure passed by IPC during module creation. For base firmware this is just some binary blob that needs to be forwarded to module. Only module knows it's structure and can consume the data. And remember that the module is external binary not linked with base firmware. 
Therefore size of this ipc_specific_config is required. Otherwise we would need to pass maximal possible configuration size which is almost 4K for IPC4 implementation. 

So to overcome this problem the proposal is to add second comp_ops definition, like this:
struct comp_dev *(*create2)(const struct comp_driver *drv,
				struct comp_ipc_config *ipc_config,
				void *ipc_specific_config, 
				uint32_t ipc_specific_config_size);

I do not see yet other option to solve this and not affect existing SOF code.

Reply:
@jxstelter if you need size for create() then we should add it to the base API (and this would save the need for a create2)

Reply:
Agree, this would be clean solution. However it will change the original comp_dev interface. I thought extending it would be more acceptable than changing its basic definition.

Reply:
Reused **struct ipc_config_process** to avoid introducing changes to comp_ops interface.